File: neet_code_150_cpp/backtracking/combination_sum_2.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

class Solution {
    public:
        // TLE
        // ビット演算で解く
        // O(N 2^N)
        std::vector<std::vector<int>> combinationSum2Bit(std::vector<int>& candidates, int target) {
            auto n = (int)candidates.size();
            std::vector<std::vector<int>> result;
            std::set<std::vector<int>> set;

            for(int bit = 0; bit < (1 << n); ++bit) {
                int sum = 0;
                std::vector<int> sub_set = {};
                for(int i = 0; i < n; ++i) {
                    if(bit & (1 << i)) {
                        sum += candidates[i];
                        sub_set.push_back(candidates[i]);
                    }
                }

                std::sort(sub_set.begin(), sub_set.end());
                if(sum == target && set.find(sub_set) == set.end()) {
                    set.insert(sub_set);
                    result.push_back(sub_set);
                }
            }

            return result;
        }
};

// 模範解答
class SolutionAns {
    public:
        std::vector<std::vector<int>> combinationSum2(std::vector<int>& candidates, int target) {
            std::sort(candidates.begin(), candidates.end());
            std::vector<int> current;
            std::vector<std::vector<int>> result;

            dfs(candidates, target, 0, 0, current, result);

            return result;
        }

    private:
        void dfs(
                std::vector<int>& candidates, 
                int target, 
                int sum, 
                int start, 
                std::vector<int>& current, 
                std::vector<std::vector<int>>& result
                ) {

            if(sum > target) {
                return;
            }
            if(sum == target) {
                result.push_back(current);
                return;
            }

            for(int i = start; i < candidates.size(); ++i) {
                if(i > start && candidates[i] == candidates[i - 1]) {
                    continue;
                }

                current.push_back(candidates[i]);
                dfs(candidates, target, sum + candidates[i], i + 1, current, result);
                current.pop_back();
            }
        }
};

void show_result(const std::vector<std::vector<int>>& result) {
    for(auto& v: result) {
        for(auto& w: v) {
            std::cout << w << " ";
        }
        std::cout << '\n';
    }
}

int main(void) {
    std::pair<std::vector<int>, int> case_1 = {{10, 1, 2, 7, 6, 1, 5}, 8};
    std::pair<std::vector<int>, int> case_2 = {{2, 5, 2, 1, 2}, 5};

    Solution s_1;

    auto res_1 = s_1.combinationSum2Bit(case_1.first, case_1.second);
    auto res_2 = s_1.combinationSum2Bit(case_2.first, case_2.second);
    show_result(res_1);
    show_result(res_2);
}

File: neet_code_150/src/backtracking/combination_sum_2.rs
// AC
// C++の模範解答より
struct SolutionAns {}
impl SolutionAns {
    pub fn combination_sum2(mut candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        candidates.sort();
        let mut current = vec![];
        let mut result = vec![];

        Self::dfs(&candidates, target, 0, 0, &mut current, &mut result);

        result
    }

    fn dfs(
        candidates: &Vec<i32>,
        target: i32,
        sum: i32,
        start: i32,
        current: &mut Vec<i32>,
        result: &mut Vec<Vec<i32>>,
    ) {
        if sum > target {
            return;
        }

        if sum == target {
            result.push(current.clone());
        }

        for i in start..candidates.len() as i32 {
            if i > start && candidates[i as usize] == candidates[i as usize - 1] {
                continue;
            }

            current.push(candidates[i as usize]);
            Self::dfs(
                candidates,
                target,
                sum + candidates[i as usize],
                i + 1,
                current,
                result,
            );
            current.pop();
        }
    }
}

fn main() {
    let case_1 = (vec![10, 1, 2, 7, 6, 1, 5], 8);
    let case_2 = (vec![2, 5, 2, 1, 2], 5);

    println!(
        "case_1: {:?}",
        SolutionAns::combination_sum2(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {:?}",
        SolutionAns::combination_sum2(case_2.0.clone(), case_2.1)
    );
}
